File: p20-29/p26.py
def MultOrd(a, n):
    k = 1
    while ((pow(a, k)-1) % n) > 0:
        k = k + 1
    return k

def period(n):
    # wikipedia says the 5 and 2 factors don't matter
    while((n % 5) == 0):
        n = n / 5
    while((n % 2) == 0):
        n = n / 2
    return MultOrd(10, n)

currlen = 0
currid = 0

for i in xrange(1,1000):
    templen = period(i)
    if(templen > currlen):
        currlen = templen
        currid = i

print currid

File: p20-29/p26.c
#include <stdio.h>
#include <math.h>

long int MultOrd(long int a, long int n)
{
  long int k;

  if(n == 1)
    return 1;
  
  for(k = 1; ((long int)pow(a,k) % n) != 1; k++)
    ;

  return k;
}

int main()
{
  long int i, j, currlen, currid, templen;
  currlen = 0;
  currid = 0;

  for(i = 3; i < 1000; i++){
    j = i;
    while(!(j % 5))
      j /= 5;
    while(!(j % 2))
      j /= 2;
    //printf("%d\t%d\n", i, j);
    templen = MultOrd(10, j);
    if(templen > currlen){
      currlen = templen;
      currid = i;
      printf("%ld\t%ld\n", currlen, currid);
    }
  }

  printf("%ld\n", currid);
  printf("%ld\n", MultOrd(10, 989));

  return 0;
}
